# 0. Import Data

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv('Churn.csv')

In [3]:
X = pd.get_dummies(df.drop(['Churn', 'Customer ID'], axis=1))
y = df['Churn'].apply(lambda x: 1 if x=='Yes' else 0)

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2)

In [5]:
y_train.head()

4450    0
5941    0
4351    0
3501    0
1623    0
Name: Churn, dtype: int64

# 1. Import Dependencies

In [6]:
from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.layers import Dense
from sklearn.metrics import accuracy_score

# 2. Build and Compile Model

In [7]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_dim=len(X_train.columns)))
model.add(Dense(units=64, activation='relu'))
model.add(Dense(units=1, activation='sigmoid'))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [9]:
model.compile(loss='binary_crossentropy', optimizer='sgd', metrics=['accuracy'])

# 3. Fit, Predict and Evaluate

In [15]:
model.fit(X_train, y_train, epochs=300, batch_size=32)

Epoch 1/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7868 - loss: 0.4341
Epoch 2/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7904 - loss: 0.4316
Epoch 3/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8016 - loss: 0.4154
Epoch 4/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.8026 - loss: 0.4099
Epoch 5/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7995 - loss: 0.4187
Epoch 6/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8008 - loss: 0.4195
Epoch 7/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8026 - loss: 0.4274
Epoch 8/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7879 - loss: 0.4258
Epoch 9/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.7997 - loss: 0.4254
Epoch 10/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.7926 - loss: 0.4272
Epoch 11/300
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7972 - loss: 0.4204
Epoch 12/300
177/177 ━━━━━━━━━━━━━━━━━━━━

In [16]:
y_hat = model.predict(X_test)
y_hat = [0 if val < 0.5 else 1 for val in y_hat]

45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [17]:
accuracy_score(y_test, y_hat)

0.7118523775727467

When the epochs =200, the accuracy score is about 0.539. When the epochs increase to 300, the accuracy score increase to 0.712.

# 4. Saving and Reloading

In [18]:
model.save('tfmodel.keras')